### ISSUE TITLE
How does loss backpropagate at self-critical phase?

### ISSUE THREAD

**Issue by forence** [ruotianluo/self-critical.pytorch#84]
Hi ruotian, 
I have a question confusing me a lot when I read your code recently. In your code, when it run into self-critical phase after cross entropy, you also calculate loss as https://github.com/ruotianluo/self-critical.pytorch/blob/713c82823a30418fa13fc47f12554c66c5b65898/train.py#L142, but `sample_logprobs` has no grad  because
https://github.com/ruotianluo/self-critical.pytorch/blob/713c82823a30418fa13fc47f12554c66c5b65898/models/AttModel.py#L247  and `gen_result.data` and `torch.from_numpy(reward).float().cuda()` have no grad obviously.
So how does loss backpropagate at this phase.

**Comment by ruotianluo** [comment]
in train.py sample_max is 0, so the samplelogprob will get gradient.

**Comment by forence** [comment]
You are right, It's my negligence ...
https://github.com/ruotianluo/self-critical.pytorch/blob/713c82823a30418fa13fc47f12554c66c5b65898/models/AttModel.py#L246-L257
I still want to know, why you must detach the  `prob_prev` to calculate  `it`? In other word, `prob_prev = torch.exp(logprobs)`, is it ok?

**Comment by ruotianluo** [comment]
It's not necessary to detach actually, multinomial will detach everything.

**Comment by forence** [comment]
Thank u very much !
